enhancements/user-monitoring.md

```diff
@@ -0,0 +1,248 @@
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+This enhancement describes a new feature for OpenShift cluster observability, namely user monitoring.
+It is an extension to the existing cluster monitoring stack to enable observability of user namespaces.
+Currently, the cluster monitoring stack only reconciles OpenShift namespaces.
+
+## Motivation
+
+Currently there is no supported solution for customers to monitor their business critical applications.
+The existing cluster monitoring stack has no support for adding scrape targets in user namespaces
+as it is only available for OpenShift components scraping workloads from `openshift-*` namespaces.
+Other Red Hat products, outside the OpenShift framework components
+don’t have a monitoring solution to integrate their metrics and out-of-the-box insights into their well being.
+Currently it is difficult to achieve metering/chargeback for Red Hat products without a single source.
+
+### Goals
+
+Allow deployment and reconcilation scrape targets, recording rules, and alerting rules in user namespaces.
+This effectively allows the reconcilation of the following custom resources in user namespaces:
+
+- ServiceMonitor (https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#servicemonitor)
+- PodMonitor (https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#podmonitor)
+- PrometheusRule (https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#prometheusrule)
+
+### Non-Goals
+
+Out of scope are:
+- Definition of Dashboards.
+- Deployment of Prometheus instances in user namespaces.
+- Deployment of Alertmanager instances user namespaces.
+
+## Proposal
+
+![](user-monitoring-deployment.png)
+
+The existing cluster monitoring stack is going to be extended with additional Prometheus (P) servers
+and one additional Prometheus Operator (PO) reconciling user namespaces
+as well as the newly added Prometheus servers.
+
+The cluster Prometheus Operator (PO) continues to operate the cluster-level Prometheus.
+We will continue to use a centralized Alertmanager (A) cluster.
+
+Querying data from a single, multi-tenant interface is done using the Thanos Querier (TQ) component.
+Soft tenancy is continued to be ensured by the existing topology leveraging kube-rbac-proxy [1]
+and prom-label-proxy [2].
+
+Differences between cluster- and user-level is mainly upgrade guarantees.
+As we keep cluster-level Prometheus immutable, we can guarantee a 100% cluster upgrade success
+but that rate may be lower for any additional Prometheus server scraping user namespaces.
+There are edge cases where an automated upgrade is not possible in the face of user monitoring.
+If users declare i.e. queries in ServiceMonitors which break in newer Prometheus versions
+we most likely cannot perform an automated migration, as PromQL is too complex to achieve that.
```

Comment:
I think it's highly unlikely that we break any compatibility with PromQL, also argument " as PromQL is too complex to achieve that." is not very convincing to me (: Can we change to maybe:

 ```suggestion
we most likely cannot perform an automated migration. Such migration without knowledge of user intention behind each query would be exceptionally difficult.
```

Reply:
Generally agreed, but for example we dropped a few PromQL functions in 1.x -> 2.x. A migration would have been impossible, as the functions simply didn't exist anymore and beyond that, modifying user controlled configuration that they probably keep in git repos sounds like a bad idea to begin with.

Reply:
Definitely, but also I think there are way more disruptive changes in those major versions updates (e.g storage format!) then just PromQL. Maybe worth to mention that?

Reply:
@bwplotka storage format is something that is under supervision of openshift namespaces, not user namespaces, hence under "tight" control. If upstream breaks that, we will most likely need to implement automatic migrations into cluster monitoring operator and/or prometheus operator.